File: presto-turbonium/src/main/java/com/facebook/presto/plugin/turbonium/storage/Values.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.presto.plugin.turbonium.storage;

import org.openjdk.jol.info.ClassLayout;

import java.util.BitSet;

import static com.facebook.presto.plugin.turbonium.storage.Util.sizeOfBitSet;
import static io.airlift.slice.SizeOf.sizeOf;

public interface Values
{
    default long getLong(int position)
    {
        throw new UnsupportedOperationException();
    }

    default int getInt(int position)
    {
        throw new UnsupportedOperationException();
    }

    default short getShort(int position)
    {
        throw new UnsupportedOperationException();
    }

    default byte getByte(int position)
    {
        throw new UnsupportedOperationException();
    }

    default boolean getBoolean(int position)
    {
        throw new UnsupportedOperationException();
    }

    default long getSizeBytes()
    {
        throw new UnsupportedOperationException();
    }

    class IntValues
            implements Values
    {
        private static final int INSTANCE_SIZE = ClassLayout.parseClass(IntValues.class).instanceSize();
        private final int[] values;

        public IntValues(int[] values)
        {
            this.values = values;
        }

        @Override
        public long getLong(int position)
        {
            return 0xffff_ffffL & values[position];
        }

        @Override
        public long getSizeBytes()
        {
            return INSTANCE_SIZE + sizeOf(values);
        }
    }

    class ShortValues
            implements Values
    {
        private static final int INSTANCE_SIZE = ClassLayout.parseClass(ShortValues.class).instanceSize();
        private final short[] values;

        public ShortValues(short[] values)
        {
            this.values = values;
        }

        public long getLong(int position)
        {
            return getInt(position);
        }

        public int getInt(int position)
        {
            return 0xffff & values[position];
        }

        @Override
        public long getSizeBytes()
        {
            return INSTANCE_SIZE + sizeOf(values);
        }
    }

    class ByteValues
            implements Values
    {
        private static final int INSTANCE_SIZE = ClassLayout.parseClass(ShortValues.class).instanceSize();
        private final byte[] values;

        public ByteValues(byte[] values)
        {
            this.values = values;
        }

        public long getLong(int position)
        {
            return getShort(position);
        }

        public int getInt(int position)
        {
            return getShort(position);
        }

        public short getShort(int position)
        {
            return (short) (0xff & values[position]);
        }
        @Override
        public long getSizeBytes()
        {
            return INSTANCE_SIZE + sizeOf(values);
        }
    }

    class BooleanValues
            implements Values
    {
        private static final int INSTANCE_SIZE = ClassLayout.parseClass(ShortValues.class).instanceSize();
        private final BitSet values;

        public BooleanValues(BitSet values)
        {
            this.values = values;
        }

        public long getLong(int position)
        {
            return getByte(position);
        }

        public int getInt(int position)
        {
            return getByte(position);
        }

        public short getShort(int position)
        {
            return getByte(position);
        }

        public byte getByte(int position)
        {
            return (byte) (values.get(position) ? 1 : 0);
        }

        public boolean getBoolean(int position)
        {
            return values.get(position);
        }
        @Override
        public long getSizeBytes()
        {
            return INSTANCE_SIZE + sizeOfBitSet(values);
        }
    }
}

File: velox/exec/Values.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "velox/exec/Values.h"
#include "velox/common/testutil/TestValue.h"

using facebook::velox::common::testutil::TestValue;

namespace facebook::velox::exec {

Values::Values(
    int32_t operatorId,
    DriverCtx* driverCtx,
    std::shared_ptr<const core::ValuesNode> values)
    : SourceOperator(
          driverCtx,
          values->outputType(),
          operatorId,
          values->id(),
          "Values"),
      valueNodes_(std::move(values)),
      roundsLeft_(valueNodes_->repeatTimes()) {}

void Values::initialize() {
  Operator::initialize();
  VELOX_CHECK_NOT_NULL(valueNodes_);
  VELOX_CHECK(values_.empty());
  // Drop empty vectors. Operator::getOutput is expected to return nullptr or a
  // non-empty vector.
  values_.reserve(valueNodes_->values().size());
  for (auto& vector : valueNodes_->values()) {
    if (vector->size() > 0) {
      if (valueNodes_->isParallelizable()) {
        // If this is parallelizable, copy the values to prevent Vectors from
        // being shared across threads.  Note that the contract in ValuesNode is
        // that this should only be enabled for testing.
        values_.emplace_back(std::static_pointer_cast<RowVector>(
            vector->copyPreserveEncodings()));
      } else {
        values_.emplace_back(vector);
      }
    }
  }
  // Drop the reference on the value nodes.
  valueNodes_ = nullptr;
}

RowVectorPtr Values::getOutput() {
  TestValue::adjust("facebook::velox::exec::Values::getOutput", this);
  if (current_ >= values_.size()) {
    if (roundsLeft_ > 0) {
      --roundsLeft_;
    }
    current_ = 0;
  }

  if (isFinished()) {
    return nullptr;
  }
  return values_[current_++];
}

void Values::close() {
  current_ = values_.size();
  roundsLeft_ = 0;
}

bool Values::isFinished() {
  return roundsLeft_ == 0;
}

} // namespace facebook::velox::exec
